### ISSUE TITLE
poetry install fails on relative paths

### ISSUE THREAD

**Issue by yoav-orca** [python-poetry/poetry#4395]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: macOS 11.5 x64
- **Poetry version**: 1.1.7
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: https://gist.github.com/yoav-orca/99a73430a8a0527ef8ca5ddad529c5b7

## Issue

I have the following layout:
```
├── service1
│   ├── poetry.lock
│   └── pyproject.toml
└── utils
    ├── lib1
    │   ├── poetry.lock
    │   └── pyproject.toml
    └── lib2
        └── pyproject.toml
```

I have service1, which depends on lib2, lib2 depends on lib1 (all relative paths).
Running `poetry install` inside the `service1` directory fails with the following exception: `Directory ../lib1 does not exist`.
`poetry install -vvv`:
```
Installing dependencies from lock file

  Stack trace:

  9  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/clikit/console_application.py:131 in run
      129│             parsed_args = resolved_command.args
      130│
    → 131│             status_code = command.handle(parsed_args, io)
      132│         except KeyboardInterrupt:
      133│             status_code = 1

  8  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/clikit/api/command/command.py:120 in handle
      118│     def handle(self, args, io):  # type: (Args, IO) -> int
      119│         try:
    → 120│             status_code = self._do_handle(args, io)
      121│         except KeyboardInterrupt:
      122│             if io.is_debug():

  7  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/clikit/api/command/command.py:171 in _do_handle
      169│         handler_method = self._config.handler_method
      170│
    → 171│         return getattr(handler, handler_method)(args, io, self)
      172│
      173│     def __repr__(self):  # type: () -> str

  6  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/cleo/commands/command.py:92 in wrap_handle
       90│         self._command = command
       91│
    →  92│         return self.handle()
       93│
       94│     def handle(self):  # type: () -> Optional[int]

  5  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/poetry/console/commands/install.py:71 in handle
       69│         self._installer.verbose(self._io.is_verbose())
       70│
    →  71│         return_code = self._installer.run()
       72│
       73│         if return_code != 0:

  4  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/poetry/installation/installer.py:103 in run
      101│         local_repo = Repository()
      102│
    → 103│         return self._do_install(local_repo)
      104│
      105│     def dry_run(self, dry_run=True):  # type: (bool) -> Installer

  3  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/poetry/installation/installer.py:239 in _do_install
      237│             self._io.write_line("Installing dependencies from lock file")
      238│
    → 239│             locked_repository = self._locker.locked_repository(True)
      240│
      241│             if not self._locker.is_fresh():

  2  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/poetry/packages/locker.py:188 in locked_repository
      186│
      187│                 package.add_dependency(
    → 188│                     Factory.create_dependency(
      189│                         dep_name, constraint, root_dir=self._lock.path.parent
      190│                     )

  1  ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/poetry/core/factory.py:244 in create_dependency
      242│                     )
      243│                 else:
    → 244│                     dependency = DirectoryDependency(
      245│                         name,
      246│                         path,

  ValueError

  Directory ../lib1 does not exist

  at ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/poetry/core/packages/directory_dependency.py:41 in __init__
       37│         self._develop = develop
       38│         self._supports_poetry = False
       39│
       40│         if not self._full_path.exists():
    →  41│             raise ValueError("Directory {} does not exist".format(self._path))
       42│
       43│         if self._full_path.is_file():
       44│             raise ValueError("{} is a file, expected a directory".format(self._path))
```

If the lock file doesn't exist, poetry install doesn't fail (strangely), but after the lock file is created, running `poetry install` again fails.

It seems that the `service1/poetry.lock` has a wrong reference to `lib1`, `lib1 = {path = "../lib1", develop = true}`  which is taken as is from `utils/lib2/poetry.lock`.

Possible Solution: 
When creating path dependencies in `[package.dependencies]`, don't copy the path (`"../lib1"`) as is from the dependency lock file (`utils/lib2/poetry.lock`). Use the uri defined in `[package.source]` (`"../utils/lib2"`) and combine it with the path (`"../lib1"`) to it. 
So the combined path is `"../utils/lib2/../lib1"`, and after compaction `"../utils/lib1"`.
In other words, we have a relative path to the dependency, and a relative path to the dependency, so we can combine them to construct a valid path relative to the current lock file.

**Comment by yoav-orca** [comment]
Closing this issue, it appears to be resolved in poetry 1.1.8!
